### ISSUE TITLE
Хотелось бы знать об ограничениях API до того как тебя забанят. Поддержка не в курсе

### ISSUE THREAD

**Issue by nleo** [TinkoffCreditSystems/invest-openapi#313]
![photo_2020-08-05_14-21-44](https://user-images.githubusercontent.com/800970/89407391-61a32680-d727-11ea-962a-7e0cca772888.jpg)
![photo_2020-08-05_14-22-38](https://user-images.githubusercontent.com/800970/89407395-636cea00-d727-11ea-8049-c270c2a0b1bb.jpg)
![photo_2020-08-05_14-22-52](https://user-images.githubusercontent.com/800970/89407398-649e1700-d727-11ea-977d-d046a0d9f67c.jpg)
![image](https://user-images.githubusercontent.com/800970/89407470-8bf4e400-d727-11ea-8a74-023123d123cf.png)

**Comment by nleo** [comment]
Т.е. проблема в том, что о всём том, что написанно в скринах люди могут узнать только после того как нарушат правила. Поддержка всегда заявляет что ограничений никаких нет и можно выставлять сколько угодно и каких угодно заявок.

**Comment by NikitaMelnikov** [comment]
Добрый день, 

Спасибо за фидбек, добавим у себя описание ограничений на бирже.
Подскажите, а по TEUR, TUSD вы через OpenApi выставляли заявки?

**Comment by mlshvdv** [comment]
@nleo а расскажите, пожалуйста, подробнее о ситуации? Чтобы сообществу иметь в виду с какими ограничениями и в следствии чего вы столкнулись )

**Comment by yurvonScreamo** [comment]
> @nleo а расскажите, пожалуйста, подробнее о ситуации? Чтобы сообществу иметь в виду с какими ограничениями и в следствии чего вы столкнулись )

неприличное большое количество заявок, вот из-за чего это произошло (10-14к в день по одной и той же бумаге)

**Comment by Fatal1ty** [comment]
> Спасибо за фидбек, добавим у себя описание ограничений на бирже.

Можете назвать сроки добавления описания ограничений?

**Comment by ProgrammingLife** [comment]
Ещё формулировка странная какая-то

> Среднее допустимое количество действий в секунду ... равное 150 на протяжении 5 минут

Это сколько же действий в секунду? Т.е. можно 150 действий отшпарить в последнюю секунду из 5 минут или столько действий не поместится в "Среднее допустимое количество действий в секунду"?

**Comment by NikitaMelnikov** [comment]
@Fatal1ty на этой неделе

**Comment by NikitaMelnikov** [comment]
@ProgrammingLife в последнюю секунду/минуту не получится из-за ограничений OpenApi

**Comment by betslus1** [comment]
> > @nleo а расскажите, пожалуйста, подробнее о ситуации? Чтобы сообществу иметь в виду с какими ограничениями и в следствии чего вы столкнулись )
> 
> неприличное большое количество заявок, вот из-за чего это произошло (10-14к в день по одной и той же бумаге)

По одной бумаге или по разным бумагам?

**Comment by felixrap** [comment]
> > @nleo а расскажите, пожалуйста, подробнее о ситуации? Чтобы сообществу иметь в виду с какими ограничениями и в следствии чего вы столкнулись )
> 
> неприличное большое количество заявок, вот из-за чего это произошло (10-14к в день по одной и той же бумаге)

Выходит, что на минутах торговать тут не получится? Даже не учитывая комиссию Тинькова. А если 120 запросов на 120 разных бумаг, получается что за сессию можно сделать 50к+ запросов.

**Comment by nonamegithub** [comment]
Дык вроде все ясно из фрагментов представленной переписки. Комиссия со сделки выходила меньше 0,01 и не могла быть удержана с него брокером, соответственно. Он и поверил в себя - решил, что теперь он самый умный человек на планете. За это и забанили.

**Comment by IvanAgafonov** [comment]
@nonamegithub Как такое получилось? Он ведь торговал TUSD и TEUR, а там не может быть такой маленькой комиссии даже при минимальном лоте, или нет? Я в курсе, что сейчас много домашних инвесторов, которые гоняют фонды от тинькофф по несколько миллионов штук за день, потому что там нету комиссии и слабое движение, но в этих фондах тогда вообще не будет никакой комиссии за сделки.

**Comment by betslus1** [comment]
Мне всегда казалось комиссия округляется вверх?
т.е. 0.000001 по факту будет 0.01? (кроме фондов Т)

**Comment by NikitaMelnikov** [comment]
TRUB, TUSD и TEUR отключены в openapi, соответственно вопрос а как торговали то собственно :)

**Comment by nonamegithub** [comment]
Currency rounding attack, классика жанра. Боюсь, не поделится товарищ с нами секретом.

**Comment by luckman** [comment]
Подозреваю, что там в округлениях дело.
Сейчас в TGLD кто-то подобным занимается.

(На картинке вверху средняя цена исполнения, хотя лимитная заявка сделана по цене 0,0812) 5 лотов стоят 0,406 округляется в 0,41)
![image](https://user-images.githubusercontent.com/2941360/91451324-fb24aa80-e885-11ea-97b6-ffb8af6b6067.png)

**Comment by nleo** [comment]
@NikitaMelnikov он торговал скриптами в браузере :-)

**Comment by nonamegithub** [comment]
О, чистосердечное признание ?!
Зачет, дружище! Я рассматривал версию с патченным .apk 
Но это тоже красиво...

**Comment by nleo** [comment]
@nonamegithub все три ситуации происходили не со мной, я просто много в каких чатах сижу, многое вижу)

Сам я обычного бота без HFT планирую

**Comment by nonamegithub** [comment]
С существующими ограничениями Tinkoff OpenAPI по количеству запросов orders в 120/мин об HFT можно забыть раз и навсегда. Следовательно на ситуацию из скриншота № 1 попасть практически нереально. Парни писали, что ловили  429 Error: Too Many Requests...

**Comment by felixrap** [comment]
> С существующими ограничениями Tinkoff OpenAPI по количеству запросов orders в 120/мин об HFT можно забыть раз и навсегда. Следовательно на ситуацию из скриншота № 1 попасть практически нереально. Парни писали, что ловили 429 Error: Too Many Requests...

Вы ещё не учитываете комиссию. Если взять тот же сбер, то, чтобы по сделке выйти в ноль и отбить комиссию (0.0025%), нужно выиграть разницу в цене копеек 12-14. И только потом, всё что выше этого - будет вашим профитом.

**Comment by NikitaMelnikov** [comment]
@nonamegithub не только ограничениями openapi, но еще и биржи

**Comment by nonamegithub** [comment]
А вот с этого момента поподробнее, пожалуйста. Бирже MOEX не нравятся запросы со средней скоростью 150 шт./сек на протяжении 5-ти минут (см. 1 скриншот). Первым и единственным заслоном в таком случае будет только Tinkoff Api со своим ограничением в 120 заявок в минуту. Нет, разве?

**Comment by Fatal1ty** [comment]
> Бирже MOEX не нравятся запросы со средней скоростью 150 шт./сек на протяжении 5-ти минут

Как я понял из [документации](https://tinkoffcreditsystems.github.io/invest-openapi/rest/), там ограничение 150 запросов в 5 минут, то есть, если делать rate limiter по минутам, то выходит 30 запросов в секунду. Так ведь?

**Comment by nonamegithub** [comment]
Нет, не так)  Цитирую дословно из скриншота № 1: "среднее допустимое количество действий в секунду... равное 150 на протяжении 5 минут".  Это, в принципе, вполне похоже на правду, Не удивлюсь, если парни с Тинькова, что-то сами не так поняли, и зарезали нас по самое не хочу...  Не скажу про заявки, но в асинхроне спокойно могу запросить котировки десятка инструментов за секунду и менее, даже через Tinkoff Api. Главное за лимит в 120/мин не выйти...

**Comment by nleo** [comment]
Я тоже по MOEX читаю, что можно сделать 5*60*150=45000 действий за пять минут, хоть все 45000 в первую минуту, или может даже в первую секунду?)

Коряво конечно они написали

**Comment by nonamegithub** [comment]
Да и [документация](https://tinkoffcreditsystems.github.io/invest-openapi/rest/) от Tinkoff подтверждает то же самое: Среднее допустимое количество действий в секунду 150 запросов в течение 5 минут.
Давайте требовать увеличения лимитов на порядок)
